### ISSUE TITLE
Is it intended for the json defaultToEncodable to be double-encoded?

### ISSUE THREAD

**Issue by sj0000e** [dart-lang/sdk#54842]
jsonEncode(myClass)
jsonEncode(jsonEncode(myClass.toMap())) for the map.
That's why I always use customToEncodable. This produces nested json, just like javascript.
jsonEncode(myClass,toEncodable: (dynamic object) => object.toMap()); 

because it's too inconvenient every time.
I wondered if defaultToEncodable using toJson was the intended behavior. Is it for security purposes?

**Comment by lrhn** [comment]
I'm not entirely sure what you are asking.

What are you doing, precisely? How does it work, and did you expect it to do something else?

**Comment by sj0000e** [comment]
javascript(Expected)
```javascript
IdMap ={'id1':new Person('John',30),'id2':new Person('Jane',31)};
JSON.stringify(IdMap);
>>> '{"id1":{"name":"John","age":30},"id2":{"name":"Jane","age":31}}'
```

dart //deafult object.toJson()
```dart
Map<String,Person> idMap ={'id1':Person('John',30),'id2':Person('Jane',31)};
print(json.encode(idMap));
>>> '{"id1":"{\"name\":\"John\",\"age\":30}","id2":"{\"name\":\"Jane\",\"age\":31}"}'
```

customToEncodable
```dart
json.encode(idMap,toEncodable:(dynamic object) => object.toMap());
>>> '{"id1":{"name":"John","age":30},"id2":{"name":"Jane","age":31}}'
```

I'm using customToEncodable in Dart because the behavior of json.encode is different from JavaScript.
But if Dart's behavior is intentional for security reasons, I need to change my code.
So, I've become curious about the intention behind using toJson.

**Comment by lrhn** [comment]
I don't know how the `Person` class is declared. If it has a `toJson` method which returns a *string*, then it's not compatible with the default fallback functionatlity of `jsonEncode`, which calls `toJson` to get an object which is JSON encoded *instead* of the un-encodable original object.
That's also what `customToEncodable` does, returns an object which is encoded *instead* of the original object.

So if `toJson` returns a string instead of a map, then that string is encoded as a JSON string, not included verbatim in the JSON encoding (after all, `jsonEncode` dosn't known that the string contains valid JSON, and it doesn't have to, a reasonable use of `toJson` would be a `Date` object returning an ISO-standard string representation of the data.)

And if the `toEncodable` function calls `toMap` instead of `toJson`, then it behaves differently.

If that's the case, that is:
```dart
class Person { 
  // ... 
  String toJson() => jsonEncode(toMap());
  Map<String, Object?> toMap() => {...fields as value...};
}
```
then the behavior you see is expected and intended. You just happen to have a `toJson` method which gets called by the default `jsonEncode` behavior, because it's there and has the right name, but which isn't actually *compatible* with that use.
So you *need* to use a custom `toEncodable` for your objects, just as you would if they had no `toJson` method at all. The objects are not compatible with the default behavior.

(In practice, I recommend *always* using a `toEncodable` function if you have any non-plain-JSON-object in the structure you want to encode. The `toJson` fallback was a bad idea, copied from JavaScript, which never really fit into the more statically typed Dart. I hope we'll eventually get rid of it entirely, so accidents like this won't happen.)

**Comment by sj0000e** [comment]
If the class has implemented toMap(), would it be appropriate to use it in json.encode()?
I believe the existing code will continue to function as is (Excluding classes that only have toMap() without toJson()),
toJson() will also be used for its original purpose.

```dart
//writeObject
if (writeJsonValue(object)) return;
if (writeJsonValue(_toEncodable(object))) return //pseudo code
if (writeJsonValue(object.toMap())) return; // add this line
throw e;
```

**Comment by lrhn** [comment]
We have no plan to add support for something like `toMap` to the Dart JSON encoder.
**The JSON encoder is not intended as a JSON serializer** (emphasized, because that's the important take-away).

There is no *standard* for encoding arbitrary objects as JSON, and even if there was a standard for JSON serialization,
the JSON encoder isn't trying to be that.
It's intended to be able to take the result of `jsonDecode`, or something of the same shape as what `jsonDecode` can create, and convert it back to a string. It encodes JSON-shaped object structures, not any object.

I'd be happy to remove support for `.toJson()` and even `toEncodable`. The goal if `jsonDecode`  and `jsonEncode` is to read and write *plain* JSON.
Anything else should be handled by code which builds, modifiers, or interprets such plain JSON object structures, *before* encoding them, not during encoding.


But, for now, we do support `toEncodable`, so you can implement whatever you want to do with objects that are not list/map/primitive values. Like a custom `toEncodable`:
```dart
Object? _cleverEncodable(Object? object) {
  try {
    var toJson = (object as dynamic).toJson();
    if (toJson is String) {
      try {
        var jsonObject = jsonDecode(toJson);
        // It was encoded JSON. Success.
        return jsonObject;
      } on Object {
        // Just a normal string.
      }
    }
    return toJson;
  } on Object {
    // Try something else.
  }
  try {
    var toMap = (object as dynamic).toMap();
    if (toMap is Map<String, Object?>) return toMap;
  } on Object {
    // Try something else.
  }
  /// Desperate fallback.
  return object.toString();
}
```
If you use that as `toEncodable`, it will try to call both `toJson` and `toMap`, and if `toJson` returns a string, it'll even check if that's possibly already encoded JSON. (Probably with some false positives, use with care.)

**Comment by sj0000e** [comment]
toJson() is intended for types like Date. 
However, I have observed a widespread desire for toJson() to return a Map.

I do not wish to discuss specific function specifications. 
Whether it's using _cleverEncodable as _defaultToEncodable or renaming toJson() to toJsonEncodable(), whatever the method, I believe that the ultimate goal is to make the Dart language user-friendly and consistent.

Your json.encode() sounds like it would make users more uncomfortable.

**Comment by sj0000e** [comment]
feel like the current code is the worst, and my alternative code is only temporary. I think your idea is the most solid. However, you haven't mentioned any plans, and the thought of this situation persisting for another 5 years is what troubles me the most. Do you have any plans for it to be revised?

**Comment by lrhn** [comment]
TL;DR: We want you tp build on top of `jsonEncode`/`jsonDecode` (or some other JSON processor), instead of trying to build things *into* it. Because what people want built into it is a long list of slightly different features, and implementing all of them will create a monster.

We have no plans to revise the current JSON encode/decode functionality. If we did, it would likely be by deprecating and removing built-in support for `toJson`, so you'd have to pass a `toEncodable` function to get that.

JSON is a text format. It can be used in many ways. It's not the goal of the Dart SDK JSON encode/decode functionality to add extra support for particular ways of using it, rather to give a consistent and correct decoding of JSON strings into similarly structured `Map`/`List`/primitive value structures, and back to strings.

Any use of JSON should build *on top* of that, not be built *in* to it.

Trying to do more things *while* encoding or decoding usually means interleaving some processing after building each map or list, but anything you can do there, you can *also* do in a post-processing pass on the created object structure, without needing to complicated the encoding and decoding.

I don't say that Dart should not have a highly efficient JSON processing library, or JSON serialization protocols, I'm saying that those shouldn't be *built in* to the base JSON encoding and parsing functionality. Specialized tools require specialized implementation, and it's not a good idea to put specialized extra code into the base operations. That usually just make both normal and specialized cases slower than they need to.

It's quite possible that a ore efficient JSON processing tool will not build on top of `jsonEncode` or `jsonDecode` at all, because there is a performance penalty in creating tall these map and list objects, if they're going to be passed to a `fromJson` constructor and deconstructed again immediately after. High performance JSON processing will likely work directly on the source, not creating intermediate objects at all. Which is another reason trying to add more functionality to `jsonEncode`/`jsonDecode` will likely are not help the JSON uses that matters most. (At least in the long run.)

**Comment by sj0000e** [comment]
I'm a little embarrassed,I've been too focused on my own thoughts.
Stepping back for a moment, I realize that I'm arguing for toJsonEncodable, while others may argue for toJson...

Thank you for your detailed explanation
